Simplenote/src/main/java/com/automattic/simplenote/models/Tag.java

```diff
@@ -92,10 +92,11 @@ public void renameTo(String tagOld, String tagNew, int index, Bucket<Note> notes
 
                 // iterate and do a case insensitive comparison on each tag
                 for (String tag : tags) {
-                    // if it's this tag, add the new tag
-                    if (getSimperiumKey().equals(TagUtils.hashTag(tag))) {
+                    // Add new tag if it's this tag and note doesn't already have same tag.
+                    if (getSimperiumKey().equals(TagUtils.hashTag(tag)) && !tags.contains(tagNew)) {
```

Comment:
we'll need to hash-compare against `tags` too won't we?

rename `cat` to `dog` but a note already has `Dog` as a tag…I think we'll still get a duplicate

```java
String thisHash = getSimperiumKey();

…
List<String> tags = new ArrayList<String>(note.getTags());
List<String> tagHashes = new ArrayList<String>();
Set<String> tagSet = new HashSet<String>();
for (String tag : tags) {
	String hash = TagUtils.hashTag(tag);
	tagHashes.add(hash);
	tagSet.add(hash);
}

List<String> newTags = new ArrayList<String>();
for (int i = 0; i < tags.size(); i++) {
	if (thisHash.equals(tagHashes[i]) && !tagSet.contains(tagHashes[i])) {
		newTags.add(newTag);
	} else {
		newTags.add(tags[i]);
	}
}
```

Reply:
Yes, we do.  Good catch!  I updated the tag comparison in 8306736.

That case led me to find other bugs with the `Tag.renameTo` method.  We also need to add a condition when creating a new tag from the new name only if the hashed tag doesn't already exist (55ddf40).